torch/distributed/_composable/fsdp/_fsdp_collectives.py

```diff
@@ -197,6 +195,23 @@ def foreach_reduce_scatter_copy_in(
     torch.cat(grad_views, dim=-1, out=reduce_scatter_input.view(world_size, -1))
 
 
+def _reduce_scatter(
+    output: torch.Tensor,
+    input: torch.Tensor,
+    group: dist.ProcessGroup,
+    divide_factors: Optional[Tuple[float, float]],
+) -> None:
+    if divide_factors:
+        predivide_factor, postdivide_factor = divide_factors
+        _div_if_needed(input, predivide_factor)
+        dist.reduce_scatter_tensor(output, input, group=group)
```

Comment:
`premul_sum` looks useful here! I think we can try to use it when we have the plumbing up to Python (IIUC, not yet?).

> Also btw, can you remind me why we are using dist.* instead of funcol here?

I think it is just legacy. Is the idea that, since we are allocating a new reduce-scatter output anyway, then there is no difference (e.g. no extra copies/allocations) in eager mode whether we use `dist.reduce_scatter_tensor` vs. `funcol.reduce_scatter_tensor`?

Reply:
> I think it is just legacy. Is the idea that, since we are allocating a new reduce-scatter output anyway, then there is no difference (e.g. no extra copies/allocations) in eager mode whether we use dist.reduce_scatter_tensor vs. funcol.reduce_scatter_tensor?

Ah that's a good point. You are right, `funcol.reduce_scatter_tensor` would perform similarly, but only if we allocated the output just-in-time in eager mode.

Reply:
I think in longer term we should consider to let funcol to accept `out` argument, this is similar behave to aten ops, and allow us to use funcol inside `dist.*` collective

Reply:
Yup this is totally doable. We can have out-variants funcols in eager, which get functionalized by AOTAutograd (we need to add functionalization glue for it I think. Not sure if the recent auto functionalization change changed this) then re-inplaced in the post-grad fx pass.